hca/upload/api_client.py

```diff
@@ -1,81 +1,238 @@
-                    url=url,
-                    status=response.status_code,
-                    content=response.content))
+    ApiClient is not normally called directly, it is used by UploadArea.
+    """
+
+    def __init__(self, deployment_stage, authentication_token=None):
+        self.deployment_stage = deployment_stage
+        # self.api_url_base = self._api_url(deployment_stage=deployment_stage)
+        self.auth_token = authentication_token
+
+    # Upload Area Manipulation
+
+    def create_area(self, area_uuid):
+        """
+        Create an Upload Area
+
+        :param str area_uuid: A RFC4122-compliant ID for the upload area
+        :return: a dict of the form { "uri": "s3://<bucket_name>/<upload-area-id>/" }
+        :rtype: dict
+        :raises UploadApiException: if the an Upload Area was not created
+        """
+        response = self._make_request('post',
+                                      path="/area/{id}".format(id=area_uuid),
+                                      headers={'Api-Key': self.auth_token})
         return response.json()
 
+    def area_exists(self, area_uuid):
+        """
+        Check if an Upload Area exists
+
+        :param str area_uuid: A RFC4122-compliant ID for the upload area
+        :return: True or False
+        :rtype: bool
+        """
+        response = requests.head(self._url(path="/area/{id}".format(id=area_uuid)))
+        return response.ok
+
+    def delete_area(self, area_uuid):
+        """
+        Delete an Upload Area
+
+        :param str area_uuid: A RFC4122-compliant ID for the upload area
+        :return: True
+        :rtype: bool
+        :raises UploadApiException: if the an Upload Area was not deleted
+        """
+        self._make_request('delete', path="/area/{id}".format(id=area_uuid),
+                           headers={'Api-Key': self.auth_token})
+        return True
+
     def credentials(self, area_uuid):
-        url = "{api_url_base}/area/{uuid}/credentials".format(api_url_base=self.api_url_base, uuid=area_uuid)
-        response = requests.post(url)
-        if not response.status_code == requests.codes.created:
-            raise RuntimeError(
-                "POST {url} returned {status}, {content}".format(
-                    url=url,
-                    status=response.status_code,
-                    content=response.content))
+        """
+        Get AWS credentials required to directly upload files to Upload Area in S3
+
+        :param str area_uuid: A RFC4122-compliant ID for the upload area
+        :return: a dict containing an AWS AccessKey, SecretKey and SessionToken
+        :rtype: dict
+        :raises UploadApiException: if credentials could not be obtained
+        """
+        response = self._make_request("post", path="/area/{uuid}/credentials".format(uuid=area_uuid))
+        return response.json()
+
+    # Files
+
+    def store_file(self, area_uuid, filename, file_content, content_type):
```

Comment:
Should this have a retry?

Reply:
Many things in this library should probably have retries.  I think we should figure out which ones, and implement the retries, under a separate ticket.  I think there is enough in this one already.